File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
       unordered_map<char,int> mp;
        int l =0, r=0,n=s.size();
        int maxi = 0;
        while(r < n){
            if(mp.find(s[r]) != mp.end()){
                l = max(l,mp[s[r]]+1);
            }

            mp[s[r]] = r;
            maxi = max(maxi,r-l+1);
            r++;
        }

        return maxi;
    }
};
/*class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size(),maxLength = INT_MIN,count=0;
        if(s.empty()) return 1;
        for(int i=0; i<s.size(); i++){
            unordered_map<char,int>mp;
            for(int j=i; j<s.size(); j++){
                 if(mp.find(s[j]) == mp.end()){
                    mp[s[j]] = 1;
                 }
                 else{
                    maxLength = max(maxLength,j-i);
                    break;
                 }
                
            }
        }
        return maxLength;
    }
};*/

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0,r=0,n=s.length();
        int maxi = 0;

        HashMap<Character,Integer> hash = new HashMap<>();
        while(r < n){
            if(hash.containsKey(s.charAt(r)) ){
                l= Math.max(l,hash.get(s.charAt(r))+1);
            }

            hash.put(s.charAt(r),r);
            maxi = Math.max(maxi,r-l+1);
            r++;
        }
        return maxi;
    }
}